model/src/main/kotlin/utils/DependencyGraphBuilder.kt

```diff
@@ -385,26 +387,38 @@ private fun Collection<DependencyReference>.toGraph(): Pair<List<DependencyGraph
     val edges = mutableListOf<DependencyGraphEdge>()
     val nodeIndices = mutableMapOf<NodeKey, Int>()
 
-    fun constructGraph(dependencies: Collection<DependencyReference>) {
-        dependencies.forEach { ref ->
-            val node = DependencyGraphNode(ref.pkg, ref.fragment, ref.linkage, ref.issues)
-            if (node !in nodes) {
-                val fromIndex = nodes.size.also { nodeIndices[ref.key] = it }
-                nodes += node
-
-                constructGraph(ref.dependencies)
+    visitEach { ref ->
+        val node = DependencyGraphNode(ref.pkg, ref.fragment, ref.linkage, ref.issues)
+        if (node !in nodes) {
+            val fromIndex = nodes.size.also { nodeIndices[ref.key] = it }
+            nodes += node
 
-                ref.dependencies.forEach { dep ->
-                    edges += DependencyGraphEdge(fromIndex, nodeIndices.getValue(dep.key))
-                }
+            ref.dependencies.forEach { dep ->
+                edges += DependencyGraphEdge(fromIndex, nodeIndices.getValue(dep))
             }
         }
     }
 
-    constructGraph(this)
     return nodes.toList() to edges
 }
 
+private fun Collection<DependencyReference>.visitEach(visit: (ref: DependencyReference) -> Unit) {
+    val visited = mutableSetOf<NodeKey>()
+    val queue = LinkedList(this)
+
+    while (queue.isNotEmpty()) {
```

Comment:
Question: Instead of this queue management, would it not be more readable with `flatMap` ?

Reply:
I find it actually readable and prefer to keep it as-is. 

I'm not exactly sure what you envision with the use of `flatMap`, so if you want to have it considered further mind sharing the code you have in mind?

Reply:
My bad... I thought it would be easy as:
```
flatMap {
        it.dependencies + it
    }.forEach {
        if (it.key !in visited) {
            visit(it)
            visited += it.key
        }
    }
```
but this code goes only to depth 1.
There are some technics with `flatMap` that allow any depth, but they are less elegant than yours with `Queue`.